### PR TITLE
feat: sealing: Split PCA/PCB batches if gas used exceeds block limit

### PR DESCRIPTION
**Pull request by shrenujbansal**
## Related Issues

https://github.com/filecoin-project/lotus/issues/10612

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: storage/pipeline/precommit_batch_test.go [hunk 1]
```diff
@@ -162,6 +162,7 @@ func TestPrecommitBatcher(t *testing.T) {
 			s.EXPECT().StateNetworkVersion(gomock.Any(), gomock.Any()).Return(network.Version14, nil)
 
 			s.EXPECT().StateMinerInfo(gomock.Any(), gomock.Any(), gomock.Any()).Return(api.MinerInfo{Owner: t0123, Worker: t0123}, nil)
+			s.EXPECT().GasEstimateMessageGas(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(&types.Message{GasLimit: 100000}, nil)
```
Comment (shrenujbansal) [inline storage/pipeline/precommit_batch_test.go:165]: Add test for splitting of PCB msg

File: storage/pipeline/commit_batch_test.go [hunk 2]
```diff
@@ -201,6 +201,7 @@ func TestCommitBatcher(t *testing.T) {
 
 			if batch {
 				s.EXPECT().StateNetworkVersion(gomock.Any(), gomock.Any()).Return(network.Version13, nil)
+				s.EXPECT().GasEstimateMessageGas(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(&types.Message{GasLimit: 1000000}, nil)
```
Comment (shrenujbansal) [inline storage/pipeline/commit_batch_test.go:204]: Add test for PCA split msg

File: node/impl/full/gas.go [hunk 3]
```diff
@@ -384,10 +384,13 @@ func gasEstimateGasLimit(
 func (m *GasModule) GasEstimateMessageGas(ctx context.Context, msg *types.Message, spec *api.MessageSendSpec, _ types.TipSetKey) (*types.Message, error) {
 	if msg.GasLimit == 0 {
 		gasLimit, err := m.GasEstimateGasLimit(ctx, msg, types.EmptyTSK)
+		log.Errorf("GasEstimateMessageGas GasLimit: %f", gasLimit)
```
Comment (magik6k) [inline node/impl/full/gas.go:387]: debug logs

File: storage/pipeline/commit_batch.go [hunk 4]
```diff
@@ -264,13 +269,13 @@ func (b *CommitBatcher) maybeStartBatch(notif bool) ([]sealiface.CommitBatchRes,
 	return res, nil
 }
 
-func (b *CommitBatcher) processBatch(cfg sealiface.Config) ([]sealiface.CommitBatchRes, error) {
+func (b *CommitBatcher) processBatch(cfg sealiface.Config, sectors []abi.SectorNumber) ([]sealiface.CommitBatchRes, error) {
```
Comment (magik6k) [inline storage/pipeline/commit_batch.go:272]: Why move from b.todo?

File: storage/pipeline/commit_batch.go [hunk 5]
```diff
@@ -365,16 +377,36 @@ func (b *CommitBatcher) processBatch(cfg sealiface.Config) ([]sealiface.CommitBa
...
+	}
+
+	// If we're out of gas, split the batch in half and try again
+	if api.ErrorIsIn(err, []error{&api.ErrOutOfGas{}}) {
+		log.Warnf("CommitAggregate message ran out of gas, splitting batch in half and trying again (sectors: %d)", len(sectors))
+		mid := len(sectors) / 2
+		ret0, _ := b.processBatch(cfg, sectors[:mid])
+		ret1, _ := b.processBatch(cfg, sectors[mid:])
+
+		return append(ret0, ret1...), nil
```
Comment (magik6k) [inline storage/pipeline/commit_batch.go:407]: This doesn't retry tho, just returns the result to sectors in two batches. All of those will just retry after some wait, and get into one oversized batch again.

File: storage/pipeline/commit_batch.go [hunk 6]
```diff
@@ -234,7 +235,11 @@ func (b *CommitBatcher) maybeStartBatch(notif bool) ([]sealiface.CommitBatchRes,
 	if individual {
 		res, err = b.processIndividually(cfg)
 	} else {
-		res, err = b.processBatch(cfg)
+		var sectors []abi.SectorNumber
+		for sn := range b.todo {
+			sectors = append(sectors, sn)
+		}
+		res, err = b.processBatch(cfg, sectors)
```
Comment (magik6k) [inline storage/pipeline/commit_batch.go:242]: This probably needs to be a loop to actually retry

Comment (shrenujbansal) [reply storage/pipeline/commit_batch.go:242]: As mentioned above, i'm not actually adding retries. Just splitting the batch recursively until it gets to an appropriate size without first sending the msg

**Review by magik6k — APPROVED** [review]
Looks good, ideally we'd prove that it works in a real setup, outside of tests.

**Pull request merged**